src/modules/bulletproofs/bulletproofs_pp_norm_product_impl.h

```diff
@@ -0,0 +1,540 @@
+    secp256k1_scratch_space* scratch,
+    const unsigned char* proof,
+    size_t proof_len,
+    secp256k1_sha256* transcript,
+    secp256k1_scalar* r,
+    const secp256k1_bulletproofs_generators* g_vec,
+    size_t g_len,
+    const secp256k1_scalar* c_vec,
+    size_t c_vec_len,
+    const secp256k1_ge* commit
+) {
+    secp256k1_scalar q, v, n, l, r_inv, h_c;
+    secp256k1_scalar *es, *s_g, *s_h, *r_inv_pows;
+    secp256k1_gej res1, res2;
+    size_t i = 0, scratch_checkpoint;
+    int overflow;
+    size_t log_n = secp256k1_bulletproofs_pp_log2(g_len), log_m = secp256k1_bulletproofs_pp_log2(c_vec_len);
+    size_t n_rounds = log_n > log_m ? log_n : log_m;
+    size_t h_len = c_vec_len;
+
+    if (g_vec->n != (h_len + g_len) || (proof_len != 65 * n_rounds + 64)) {
+        return 0;
+    }
+
+    if (!secp256k1_is_power_of_two(g_len) ||  !secp256k1_is_power_of_two(h_len)) {
+        return 0;
+    }
+
+    secp256k1_scalar_set_b32(&n, &proof[n_rounds*65], &overflow); /* n */
+    if (overflow) return 0;
+    secp256k1_scalar_set_b32(&l, &proof[n_rounds*65 + 32], &overflow); /* l */
+    if (overflow) return 0;
+    if (secp256k1_scalar_is_zero(r)) return 0;
+
+    /* Compute powers of q_inv. Later used in g_factor computations*/
+    secp256k1_scalar_inverse_var(&r_inv, r);
+    scratch_checkpoint = secp256k1_scratch_checkpoint(&ctx->error_callback, scratch);
+
+    /* Collect the challenges in a new vector */
+    r_inv_pows = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, log_n * sizeof(secp256k1_scalar));
+    es = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, n_rounds * sizeof(secp256k1_scalar));
+    s_g = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, g_len * sizeof(secp256k1_scalar));
+    s_h = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, h_len * sizeof(secp256k1_scalar));
+    if (r_inv_pows == NULL || es == NULL || s_g == NULL || s_h == NULL) {
+        secp256k1_scratch_apply_checkpoint(&ctx->error_callback, scratch, scratch_checkpoint);
+        return 0;
+    }
+
+    secp256k1_bulletproofs_powers_of_r(r_inv_pows, &r_inv, log_n);
+
+    for (i = 0; i < log_n; i++) {
+        secp256k1_scalar_sqr(r, r);
+    }
+
+    for (i = 0; i < n_rounds; i++) {
+        secp256k1_scalar e;
+        secp256k1_sha256_write(transcript, &proof[i * 65], 65);
+        secp256k1_bulletproofs_challenge_scalar(&e, transcript, 0);
+        es[i] = e;
+    }
+    /* s_g[0] = n * r^(n - 1) = n * (r)^n * r_inv = n * q * r_inv */
+    secp256k1_scalar_mul(&s_g[0], &n, r);
+    secp256k1_scalar_mul(&s_g[0], &s_g[0], &r_inv);
+    for (i = 1; i < g_len; i++) {
+        size_t log_i = secp256k1_bulletproofs_pp_log2(i);
+        size_t nearest_pow_of_two = (size_t)1 << log_i;
+        secp256k1_scalar_mul(&s_g[i], &s_g[i - nearest_pow_of_two], &es[log_i]);
+        secp256k1_scalar_mul(&s_g[i], &s_g[i], &r_inv_pows[log_i]);
+    }
```

Comment:
Ok, the rust code does require a second loop. @Liam-Eagen explained to me how your code can this save one loop. Cool! But would be good to document this better.

Reply:
You added a comment a few lines, but I still couldn't make sense of the original comment: n is used both to refer to the reduced norm vector as well as the length of the original norm vector. `q` is in the equation but not used in the code below. `r` doesn't refer to the `r` argument provided to the function.

I created a new branch https://github.com/jonasnick/secp256k1-zkp/commits/b%2B%2B-jn-2 that contains an attempt to improve this and also adds a few other variable renames for clarification (e.g., `log_n` in this PR isn't the logarithm of `n` but the logarithm of the length of the norm vector).

I also made the `r` argument constant because that seemed much less error prone for the caller. But maybe this was intententional and the caller should actually get `r_f` from the function? In PR #207 it doesn't look like the `r_f` output is being used. Same for the prover, so could we make it constant for the prover as well?

Reply:
Sorry, I missed this comment and forgot to address this. Overall, I like your fixups. Will add those.

Reply:
> I also made the r argument constant because that seemed much less error prone for the caller. But maybe this was intentional and the caller should actually get r_f

I only did this because it was an internally used function and wanted to avoid the extra copy. In an earlier implementation(1~ week ago) I had everything as const `n_vec`, `l_vec`, `c_vec` as consts. 

But for avoiding the mallocs inside the `secp256k1_bpp_norm_arg_prove_impl`, I changed those to non-const versions. While doing that, I also changed r. Don't mind keeping the r as `const` if it helps to understand.

Reply:
I think there's a difference between copying O(1) (`r`) and mallocing+copying O(n) (`n_vec`, `l_vec`, `c_vec`). I think the former is definitely fine, the latter is debatable. So if that was the only reason for making `r` not const, then I think it'd be better if we'd make it `const` again.

Reply:
Force pushed to fix this.

Reply:
Done

Reply:
Done